GSA-TTS/FAC
# Character Limits - General Info & Submission Access

# Character Limits - General Info & Submission Access

Issue: https://github.com/GSA-TTS/FAC/issues/1999

Character limits for the general information and submission access fields.

See [this drive](https://drive.google.com/drive/folders/1Pb0PxOaGiky5vPbbeIdsFTnDzZGnqO9o?usp=sharing) for a quick report on average/median/min/max value lengths in dissemination. 

I pulled the min/max values from the above report into JSON, and gave them their own folder and README in `backend/schemas/character_limits`. From the initial values, I changed the values up/down to reasonable values. The README contains explanations for these decisions. Most fields in this folder remain unchanged from the report, and can be used in a later PR for setting limits on the workbooks.

## Changes:

**In this PR, I set character limits for the General information and Submission access pages.**

1. `backend/schemas/character_limits` exists. `general.json` has values edited from the baseline report.
2. The values from `general.json` are used to set min/maxLength values:
    * In `GeneralInformation.schema.json` and `Baselibsonnet`
    * In `backend/report_submission/forms.py` for the General Information form
    * In `backend/api/serializers` for the Submission access form
3. HTML changes in `gen-form.html` ensure that errors display correctly
4. JS changes in `validate.js` allow multiple length validations, and include greater/less than comparators
5. HTML changes in `step-3.html` fix a little wonkyness (not all), but also display character limits during the form entry process (as opposed to post-form submission, like in gen info).
6. Auto-formatting changes for schema files. This comes from `make all --> format_jsonnet -> jsonnetfmt -i {files}`

I've also changed the regex for `audit_period_other_months` to allow for single character responses. This resolves https://github.com/GSA-TTS/FAC/issues/3969, which has been an issue difficult to discern for users for a little while.

`^0?[1-9]$|^1[0-8]$`
`^0?[1-9]$` - One word, either 1-9 or 01-09 OR `^1[0-8]$` - One word, 10-18. 

Other validation errors now provide a generic "unexpected error" response, so these should be caught much faster in the future. This resolves https://github.com/GSA-TTS/FAC/issues/3970.

## How to test:
1. Switch to this branch and run normally. 
    * Ensure that this step rebuilds JS, so the validate.js changes are brought in
3. Navigate to the submission access page (step three, right before the report is created and general information starts)
4. Verify that character limits apply: 
    * Try to enter weird-length names (less than 2 characters, or more than 100)
    * Try to enter weird-length emails (less than 6 characters `a@a.a`, or more than 340)
7. Navigate to the general information page
8. Verify that character limits apply, and that you get nicer error messaging:
    * If a field has too many or too few characters, a nice error should come back.
    * If a field has NO characters, the form should still pass.
    * `audit_period_other_months` should allow a single digit response, and provide a nice error if it is outside the allowed range (1-18 months).

## Screenshots:

### Character limits:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/6beca2c4-068e-4da8-b7f3-a82906192f60)

![image](https://github.com/GSA-TTS/FAC/assets/91098850/17cad425-1016-45a4-ba65-829a83e2405d)

### `audit_period_other_months`

Regex issue:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/5c390749-1e22-4b4f-a6b7-643576df1c59)

Length and regex issue:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/ccd3e1d6-31c2-4817-bdbd-9941de3a44be)

## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
